[quant][graphmode][refactor] propagateQuantizationOps

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39645 [quant][graphmode] Support quantized::batch_norm2d_relu fusion for tracing
* #39644 [quant][graphmode] Support quantization for `aten::append`
* #39615 [quant][graphmode] Fix a corner case in handling `if` in insert_observers
* #39551 [quant][graphmode] Support propagate dequantize for nodes with multiple outputs
* **#39550 [quant][graphmode][refactor] propagateQuantizationOps**

Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942063](https://our.internmc.facebook.com/intern/diff/D21942063)

jerryzh168 committed:
[quant][graphmode][refactor] propagateQuantizationOps

Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -826,9 +826,7 @@ void propagateQParams(
-void propagateDequantize(Value* output, const std::vector<Value*> inputs) {
-  Node* n = output->node();
-  Graph* graph = n->owningGraph();
+void removeDequantizeFromInputs(const std::vector<Value*>& inputs) {
```
Can you add a test showing the functionality here before you land?

jerryzh168 replied:
this is a refactor, not a new functionality, it's already tested in test_quantize_script.py

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942063](https://our.internmc.facebook.com/intern/diff/D21942063)

[ghstack-poisoned]

supriyar commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -860,26 +877,23 @@ void propagateQuantizationOps(Block* block) {
-        for (auto* input : inputs) {
-          is_dequantized &= input->node()->kind() == Symbol::aten("dequantize");
+    if (isSingleInputGeneralValueAtenFunction(n)) {
+      for (auto* output : n->outputs()) {
```
Do we need this for loop inside the `if`, `else` statements? Can it be moved outside?

jerryzh168 replied:
yeah, this refactor is preparing for next PR, where we do things differently for the else branch

supriyar reviewed (approved):
LG

jerryzh168 committed:
Update on "[quant][graphmode][refactor] propagateQuantizationOps"


Summary:
This is to prepare for next PR that fixes propagate dequantize for ops with multiple outputs

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21942063](https://our.internmc.facebook.com/intern/diff/D21942063)

[ghstack-poisoned]
